lib/filter.js

```diff
@@ -0,0 +1,141 @@
+
+  var splitElement = filter._splitElement(scalarElement);
+  if (!splitElement) return { error: "empty filter" };
+
+  var error = filter._stringOnlyOperator(splitElement.operator, attributeConfig);
+  if (error) return { error: error };
+
+  if (attributeConfig._settings) {  // relationship attribute: no further validation
+    return { result: splitElement };
+  }
+
+  var validateResult = attributeConfig.validate(splitElement.value);
+  if (validateResult.error) {
+    return { error: validateResult.error.message };
+  }
+
+  var validatedElement = { operator: splitElement.operator, value: validateResult.value };
+  return { result: validatedElement };
+};
+
+filter._parseFilterElementHelper = function(attributeConfig, filterElement) {
+  if (!filterElement) return { error: "invalid or empty filter element" };
+
+  var parsedElements = [].concat(filterElement).map(function(scalarElement) {
+    return filter._parseScalarFilterElement(attributeConfig, scalarElement);
+  });
+
+  if (parsedElements.length === 1) return parsedElements[0];
+
+  var errors = parsedElements.reduce(function(combined, element) {
+    if (!combined) {
+      if (!element.error) return combined;
+      return [ element.error ];
+    }
+    return combined.concat(element.error);
+  }, null);
+
+  if (errors) return { error: errors };
+
+  var results = parsedElements.map(function(element) {
+    return element.result;
+  });
+
+  return { result: results };
+};
+
+filter._parseFilterElement = function(attributeName, attributeConfig, filterElement) {
+  var helperResult = filter._parseFilterElementHelper(attributeConfig, filterElement);
+
+  if (helperResult.error) {
+    return {
+      error: {
+        status: "403",
+        code: "EFORBIDDEN",
+        title: "Invalid filter",
+        detail: "Filter value for key '" + attributeName + "' is invalid: " + helperResult.error
+      }
+    };
+  }
+  return { result: helperResult.result };
+};
+
+filter.process = function(request, callback) {
+  if (!request.params.filter) return callback();
+
+  var resourceConfig = request.resourceConfig;
+
+  var processedFilter = { };
+  var error;
+  var filterElement;
+  var parsedFilterElement;
+
+  for (var key in request.params.filter) {
+    filterElement = request.params.filter[key];
+
+    if (!Array.isArray(filterElement) && filterElement instanceof Object) continue;  // skip deep filters
+
+    error = filter._resourceDoesNotHaveProperty(resourceConfig, key);
+    if (error) return callback(error);
+
+    error = filter._relationshipIsForeign(resourceConfig, key);
+    if (error) return callback(error);
+
+    parsedFilterElement = filter._parseFilterElement(key, resourceConfig.attributes[key], filterElement);
+    if (parsedFilterElement.error) return callback(parsedFilterElement.error);
+
+    processedFilter[key] = [].concat(parsedFilterElement.result);
+  }
+
+  request.processedFilter = processedFilter;
+
+  return callback();
```

Comment:
Agreed. I guess the refactoring from its origins as a waterfall step led in this direction. I'll make the changes.